src/coprocessor/codec/convert.rs

```diff
@@ -45,51 +86,157 @@ pub fn truncate_f64(mut f: f64, flen: u8, decimal: u8) -> Res<f64> {
 
 /// `overflow` returns an overflowed error.
 macro_rules! overflow {
-    ($val:ident, $bound:ident) => {{
-        Err(box_err!("constant {} overflows {}", $val, $bound))
+    ($val:ident, $bound:tt) => {{
+        Error::Eval(
+            format!("constant {} overflows {}", $val, $bound),
+            ERR_DATA_OUT_OF_RANGE,
+        )
     }};
 }
 
+/// `convert_int_to_int` converts an int value to a diferent int value.
+#[inline]
+pub fn convert_int_to_int(ctx: &mut EvalContext, val: i64, tp: FieldTypeTp) -> Result<i64> {
+    let lower_bound = integer_signed_lower_bound(tp);
+    // https://dev.mysql.com/doc/refman/8.0/en/out-of-range-and-overflow.html
+    if val < lower_bound {
+        ctx.handle_overflow(overflow!(val, lower_bound))?;
+        return Ok(lower_bound);
+    }
+    let upper_bound = integer_signed_upper_bound(tp);
+    if val > upper_bound {
+        ctx.handle_overflow(overflow!(val, upper_bound))?;
+        return Ok(upper_bound);
+    }
+    Ok(val)
+}
+
 /// `convert_uint_to_int` converts an uint value to an int value.
 #[inline]
-pub fn convert_uint_to_int(val: u64, upper_bound: i64, tp: FieldTypeTp) -> Result<i64> {
+pub fn convert_uint_to_int(ctx: &mut EvalContext, val: u64, tp: FieldTypeTp) -> Result<i64> {
+    let upper_bound = integer_signed_upper_bound(tp);
     if val > upper_bound as u64 {
-        return overflow!(val, tp);
+        ctx.handle_overflow(overflow!(val, upper_bound))?;
+        return Ok(upper_bound);
     }
     Ok(val as i64)
 }
 
 /// `convert_float_to_int` converts an f64 value to an i64 value.
 ///  Returns the overflow error if the value exceeds the boundary.
-pub fn convert_float_to_int(
-    fval: f64,
-    lower_bound: i64,
-    upper_bound: i64,
-    tp: FieldTypeTp,
-) -> Result<i64> {
+#[inline]
+pub fn convert_float_to_int(ctx: &mut EvalContext, fval: f64, tp: FieldTypeTp) -> Result<i64> {
     // TODO any performance problem to use round directly?
     let val = fval.round();
+    let lower_bound = integer_signed_lower_bound(tp);
     if val < lower_bound as f64 {
-        return overflow!(val, tp);
+        ctx.handle_overflow(overflow!(val, lower_bound))?;
+        return Ok(lower_bound);
     }
 
+    let upper_bound = integer_signed_upper_bound(tp);
     if val > upper_bound as f64 {
-        return overflow!(val, tp);
+        ctx.handle_overflow(overflow!(val, upper_bound))?;
+        return Ok(upper_bound);
     }
     Ok(val as i64)
 }
 
+/// `convert_bytes_to_int` converts a byte arrays to an i64 in best effort.
+pub fn convert_bytes_to_int(ctx: &mut EvalContext, bytes: &[u8], tp: FieldTypeTp) -> Result<i64> {
+    let s = str::from_utf8(bytes)?.trim();
```

Comment:
For the `?` op, is this behaviour correct when it is a binary string?

Reply:
I think invalid string should not be parsed.

Reply:
In MySQL invalid UTF-8 strings are indeed got parsed:

```
mysql> select cast(unhex('313233FF') as SIGNED);
+-----------------------------------+
| cast(unhex('313233FF') as SIGNED) |
+-----------------------------------+
|                               123 |
+-----------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+----------------------------------------------+
| Level   | Code | Message                                      |
+---------+------+----------------------------------------------+
| Warning | 1292 | Truncated incorrect INTEGER value: '123\xFF' |
+---------+------+----------------------------------------------+
1 row in set (0.00 sec)
```